server/sonar-process-monitor/src/main/java/org/sonar/process/monitor/Monitor.java

```diff
@@ -309,29 +337,60 @@ private boolean didAnyProcessRequestRestart() {
   }
 
   public class HardStopWatcherThread extends Thread {
-    private final ProcessCommands processCommands;
 
-    public HardStopWatcherThread(ProcessCommands processCommands) {
+    public HardStopWatcherThread() {
       super("Hard stop watcher");
-      this.processCommands = processCommands;
     }
 
     @Override
     public void run() {
       while (lifecycle.getState() != Lifecycle.State.STOPPED) {
-        if (processCommands.askedForStop()) {
+        if (askedForStop()) {
           trace("Stopping process");
           Monitor.this.stop();
         } else {
-          try {
-            Thread.sleep(WATCH_DELAY_MS);
-          } catch (InterruptedException ignored) {
-            // keep watching
-          }
+          delay();
         }
       }
     }
 
+    private boolean askedForStop() {
+      File tempDir = fileSystem.getTempDir();
+      try (DefaultProcessCommands processCommands = new DefaultProcessCommands(tempDir, CURRENT_PROCESS_NUMBER, false)) {
+        if (processCommands.askedForStop()) {
+          return true;
+        }
+      } catch (IllegalArgumentException e) {
+        // DefaultProcessCommand currently wraps IOException into a IllegalArgumentException
+        // accessing the shared memory file may fail if a restart is in progress and the temp directory
+        // is currently deleted and not yet recreated
+        if (e.getCause() instanceof IOException) {
+          ignore((IOException) e.getCause());
+        } else {
+          rethrow(e);
+        }
+      } catch (Exception e) {
+        rethrow(e);
+      }
+      return false;
+    }
+
+    private void ignore(IOException e) {
+      trace("HardStopWatcherThread: Error checking stop flag in shared memory. Ignoring. Keep on watching.", e);
+    }
+
+    private void rethrow(Exception e) {
+      throw new RuntimeException("Unexpected error occurred while watch for stop flag in shared memory", e);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00112)